### ISSUE TITLE
[RA2 Ch4] k8s API version requirement

### ISSUE THREAD

**Issue by rihabbanday** [cntt-n/CNTT#2132]
`ra2.k8s.005` specifies that the implementation must use Kubernetes `current release - 2 `. This introduces tight time bounds since the upstream installer project utilized by the implementation might not meet this requirement. For example, the current RI2 is three versions old because of the upstream project it utilizes. This requirement should either be relaxed, or the possibility of defining an exception should be added.

**Comment by collivier** [comment]
Anything else would be end of life.
The current requirement is currently falsy as it should pick a specific version, current release as it was discussed earlier.
It defcto asks for 3 RC as even K8s conformance (interoperability testing) works for a specific release.

**Comment by collivier** [comment]
I would consider RI2 being 3 releases late is a RI2 issue.

**Comment by rihabbanday** [comment]
@collivier The issue is that RI2 (and may be other installers in future) consumes upstream installer projects which might be behind latest k8s API releases (master). In those cases, there should be some exception. @pgoyal01 had mentioned yesterday that there is an exception process that could be used to handle such cases.

**Comment by pgoyal01** [comment]
@rihabbanday Please see [RA2 1.3.2 Exceptions](https://github.com/cntt-n/CNTT/blob/master/doc/ref_arch/kubernetes/chapters/chapter01.md#132-exceptions).

I brought up your issue at the RA2 meeting and it was discussed; 2 possible solutions are listed in the minutes:

> o	Exception to the 3-version-rule for Kubernetes support to be formulated and included - this is due to a zero-delay between community releases and the application of the rule, and network operators having slightly longer lead times to deploy.  
       - Either we include a delay to apply the rule 
       - or we have an exception process for an application to apply for. > 1.3.2.1 under ch01 - or change req.gen.cnt.03 to have a delay

**Comment by CsatariGergely** [comment]
The version requirement is inherited form Kubernetes Conformance (https://github.com/cncf/k8s-conformance/blob/master/terms-conditions/Certified_Kubernetes_Terms.md#qualifying-offerings-and-self-testing). Releasing the version requirement practically mean that we do not require a Kubernetes Conformant installation, however it is also possible to request that the installation was Kubernetes conformant a point of time in the past..

**Comment by tomkivlin** [comment]
> The version requirement is inherited form Kubernetes Conformance (https://github.com/cncf/k8s-conformance/blob/master/terms-conditions/Certified_Kubernetes_Terms.md#qualifying-offerings-and-self-testing). Releasing the version requirement practically mean that we do not require a Kubernetes Conformant installation, however it is also possible to request that the installation was Kubernetes conformant a point of time in the past..

The version requirement was actually pulled from the "supported versions" document: https://kubernetes.io/docs/setup/release/version-skew-policy/

There are two separate things here.
1. Whether or not the version of Kubernetes used is conformant - as Gergely mentions this can include older versions as each version is tested and recorded in the repo, and this spreadsheet: https://docs.google.com/spreadsheets/d/1LxSqBzjOxfGx3cmtZ4EbB_BGCxT_wlxW_xgHVVa23es/edit#gid=0.
    - e.g. kubespray 1.16 was conformant and therefore we could choose to use that, we meet the "conformant distribution" requirement
1. Whether or not the version of Kubernetes used is supported by the community and will therefore get appropriate patches etc.  This is perhaps one that is more easily relaxed as each operator can make their decision based on the installer/distribution used, their own risk position, relationship with their vendor (if applicable), their skills, etc.
    - e.g. kubespray 1.16 is conformant, but is now unlikely to receive any patches so there is increased risk to the operator - some operators may decide that is OK because they have good k8s / Go skills to fix things, other operators may not.
    - e.g. Platform Vendor A include 1.16 in their distribution, operator who is paying that vendor deems that ok despite it being n-4 because they trust Platform Vendor A to fix the problems for them as per their contract terms, or face penalties.

TL/DR:  I think it would be ok to relax this requirement, but clarify in chapter03.md the difference between the two so we continue to ensure a conformant version of Kubernetes is used, but the version used is up to the operator.

**Comment by tomkivlin** [comment]
Although that does make both RC2 harder, and the development of software - if the k8s version isn't fixed.

**Comment by CsatariGergely** [comment]
Other option is that the implementation uses an earlier CNTT RA2 release. 
Is this solution good enough to close this issue?
What do you think @rihabbanday , @tomkivlin ?

**Comment by tomkivlin** [comment]
Makes sense to me.  So e.g. RI2 release 3 aligns with RA2 release 2?

**Comment by tomkivlin** [comment]
Release process was mentioned in Anuket TSC meeting 12th Jan 2021. Meeting notes here: https://wiki.anuket.io/display/HOME/2021-01-12+TSC+Meeting.

Presentation by David McBride here: https://wiki.anuket.io/download/attachments/67776/Anuket%20Release%20Process%20Issues%20and%20Objectives.pdf?version=1&modificationDate=1610460879783&api=v2

**Comment by rihabbanday** [comment]
We discussed this during todays RI2 meeting - the new release plan seems reasonable and solves the versioning issue. I will create a PR in RI2 to reflect this information. We can close this issue if there are no further comments.

**Comment by tomkivlin** [comment]
@rihabbanday I am happy for the issue to be closed.

**Comment by tomkivlin** [comment]
Release strategy to be discussed at vF2F event.
